### PR TITLE
[sbt-gen] sbt plugin for openapi to endpoints codegen

### LINKED ISSUE #3180: Create a OpenAPI to Endpoint Code Gen sbt plugin
- Needs to have a config to switch between automatic (on file change) and manual generation
- config for file/folder path(s) of OpenAPI files
- config for package prefix
- setting to handover a global openapi config
- setting for openapi config based on a sbt file filter

### PR DESCRIPTION
**Pull request by hochgi**
/fixes #3180
/claim #3180

* based on a plugin I collaborated with @runtologist on, in another closed repo
* automatic mode integrated in regular `~compile` continuous compilation mode, manual trigger via direct task call
* static openapi files read from `src/main/oapi/**`
* dynamic download/create/… of openapi files via regular use of `sourceGenerators` scoped to the custom `ZIOpenApi` configuration. Files should reside under `ZIOpenApi / sourceManaged` (which is by default evaluated as `target/scala-{cross version}/src_managed/oapi/` in same dir as package convention (see example in tests).
* package prefix determined by relative path under `/src/main/oapi/` => allows for multiple different spec files in different packages in a conventional and expected manner
* sbt config to control codegen
* using sbt caching mechanism to avoid re-generating sources from unchanged spec files on every compile

known gaps:
* for some (currently unknown) reason, using `.dependsOn(…)` does not work, and I was forced to use `zio-http-gen` as an external dependency
* cross compilation fails, currently commented out, so we only support sbt 1.x (sbt 2 is not yet released anyway)
* ~~only "static" openapi spec files are supported as of this first commit. Ideally, we should allow for dynamic generation of spec files like any other generated sources.~~ (done!)
* collisions of generated sources from different spec files is not checked (but users should really use different packages for different services they're generating endpoints for).
* only supporting `Compile` (main) configuration - generating endpoints  for e.g. `Test` (by placing openapi files under `src/test/oapi/`) will not work out of the box (but is it really needed anyway?)
* no README/docs yet - should probably add to the site, and document after we converge on actual functionality (I assume review might change enough to render any docs I'll write now obsolete)

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (hochgi) [inline .github/workflows/ci.yml:93]: @987Nabil this was naive, not sure how to add sbt plugin testing to CI.
this does not work:
```
[error] java.lang.RuntimeException: ci.yml does not contain contents that would have been generated by sbt-github-actions; try running githubWorkflowGenerate
[error] 	at scala.sys.package$.error(package.scala:30)
[error] 	at sbtghactions.GenerativePlugin$.reportMismatch$1(GenerativePlugin.scala:919)
[error] 	at sbtghactions.GenerativePlugin$.compare$1(GenerativePlugin.scala:925)
[error] 	at sbtghactions.GenerativePlugin$.$anonfun$projectSettings$5(GenerativePlugin.scala:929)
[error] 	at sbtghactions.GenerativePlugin$.$anonfun$projectSettings$5$adapted(GenerativePlugin.scala:906)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:63)
[error] 	at sbt.std.Transform$$anon$4.work(Transform.scala:69)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:283)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:24)
[error] 	at sbt.Execute.work(Execute.scala:292)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:283)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:65)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
[error] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
[error] 	at java.base/java.lang.Thread.run(Thread.java:1583)
[error] (githubWorkflowCheck) ci.yml does not contain contents that would have been generated by sbt-github-actions; try running githubWorkflowGenerate
```

How can I add a CI step to run `sbt zioHttpGenSbt/scripted`?

File: zio-http-gen-sbt-plugin/src/main/scala/zio/http/gen/sbt/Format.scala
Comment (hochgi) [inline zio-http-gen-sbt-plugin/src/main/scala/zio/http/gen/sbt/Format.scala:10]: pretty naive - if needed, this can be enhanced by simply trying to parse as json, and then yaml. or even use apache-tika to detect file's mimetype.
For now, this seems good enough to me.

File: build.sbt
Comment (hochgi) [inline build.sbt]: wasteful.
I wanted to add sbt testing to CI, and this works. but…
it has scalaVersion set to 2.12 (only supporting sbt 1.x for now),
and it runs for every matrix value. which means we run the same tests multiple times.
Not sure how to opt out.

File: build.sbt
Comment (hochgi) [inline build.sbt]: this does not work in CI.
Since sbt 2.x is not published yet (only milestone versions currently available),
I commented it out.
When sbt 2.0 comes out, this should be commented in with stable version, and it should just work.

File: build.sbt
Comment (hochgi) [inline build.sbt]: I tried, and wasted a few hours trying to figure out why it won't work.
In a clean multi module project, it works.
Only in zio-http repo it acts weird.

similar plugin modules as part of a larger build work in other repos (playframework, tapir, …)

Any idea what's special here?

File: zio-http-gen-sbt-plugin/src/main/scala/zio/http/gen/sbt/ZioHttpCodegen.scala
Comment (hochgi) [inline zio-http-gen-sbt-plugin/src/main/scala/zio/http/gen/sbt/ZioHttpCodegen.scala]: nit: can be tail recursive
```scala
@tailrec def listFilesRec(dir: File, acc: List[File] = Nil): List[File] =
  sbt.io.IO.listFiles(dir).foldRight(acc) { case (f, tail) =>
    if (f.isDirectory) listFilesRec(f, tail)
    else f :: tail
  }
```

File: zio-http-gen-sbt-plugin/src/sbt-test/zio-http-sbt-codegen/dynamic/build.sbt
Comment (hochgi) [inline zio-http-gen-sbt-plugin/src/sbt-test/zio-http-sbt-codegen/dynamic/build.sbt:19]: arbitrary openapi spec. I tried so many, and every single one I checked had usage of features we don't support in zio-http-gen yet.
not the official swagger example: https://petstore3.swagger.io
nothing in this collection (didn't do an exhaustive check though): https://petstore.dev
So I tried looking for the most minimal example I could find, and even the one I found didn't work out of the box.
So I pruned the JSON, to only be left with the acceptable APIs definitions. 

In retrospect, it kind of nice to have this as an example.
Most services we deal with have many APIs, and we don't use all of them.
So no need to generate code for stuff we're not using.

I may have came up with this example out of necessity, but now I think it's actually good to have it.

Comment (hochgi) [reply zio-http-gen-sbt-plugin/src/sbt-test/zio-http-sbt-codegen/dynamic/build.sbt:19]: Also, I don't like the idea of tests depending on stuff outside of the build (other than conventional stuff like library dependencies of course)

But this is how I expect users to actually use it. so… ¯\\\_(ツ)\_/¯

File: zio-http-gen-sbt-plugin/src/sbt-test/zio-http-sbt-codegen/dynamic/test
Comment (hochgi) [inline zio-http-gen-sbt-plugin/src/sbt-test/zio-http-sbt-codegen/dynamic/test]: nit:
might want to also check for existence of `target/scala-2.13/src_managed/oapi/gov/uspto/ibd/api.json`
Just so it can serve as documentation for users where to expect dynamic openapi files to be found.

File: build.sbt
Comment (987Nabil) [reply build.sbt]: Why does it not work with 2 M2? I saw other plugins having cross-builds
If it does not, please put this code in an issue, not in a comment.

File: build.sbt
Comment (987Nabil) [reply build.sbt]: idk
But this is no option, if the plugin is in the same repo. I actually started myself to work on this issue and and problems with the sbt setup. So I created a different repo. Maybe we should do that.

File: zio-http-gen-sbt-plugin/src/main/scala/zio/http/gen/sbt/Format.scala
Comment (987Nabil) [reply zio-http-gen-sbt-plugin/src/main/scala/zio/http/gen/sbt/Format.scala:10]: okay for now

File: zio-http-gen-sbt-plugin/src/main/scala/zio/http/gen/sbt/ZioHttpCodegen.scala
Comment (987Nabil) [reply zio-http-gen-sbt-plugin/src/main/scala/zio/http/gen/sbt/ZioHttpCodegen.scala]: this methods should be private?

File: build.sbt
Comment (hochgi) [reply build.sbt]: Its solveable. same structure works in playframework and tapir projects.
It's also easier to catch integration issues.
If something changes in the `zio-http-gen` module which breaks the plugin, ideally we would want to catch that right away.

I'll try to spend some more time on this :/

File: zio-http-gen-sbt-plugin/src/main/scala/zio/http/gen/sbt/ZioHttpCodegen.scala
Comment (hochgi) [reply zio-http-gen-sbt-plugin/src/main/scala/zio/http/gen/sbt/ZioHttpCodegen.scala]: I don't have any strong opinions about it, if you think such stateless methods in plugins are best to limit access to, then OK, I guess.

File: build.sbt
Comment (hochgi) [reply build.sbt]: I don't know why.
maybe those versions are published to another repo, which we need to add another resolver for?
might be as simple as that, I didn't checked.

an issue would be best to put such TODOs

File: zio-http-gen-sbt-plugin/src/main/scala/zio/http/gen/sbt/ZioHttpCodegen.scala
Comment (987Nabil) [reply zio-http-gen-sbt-plugin/src/main/scala/zio/http/gen/sbt/ZioHttpCodegen.scala]: This is not a plugin to recurse through files

**Comment by 987Nabil** [comment]
Do I see it right, that compile always triggers the generation? I'd like it to be opt out

File: build.sbt
Comment (987Nabil) [reply build.sbt]: M2 is on maven central. See https://mvnrepository.com/artifact/org.scala-sbt/sbt

Comment (hochgi) [reply build.sbt]: I'll open an issue then.
should I wait until this is merged?

**Comment by hochgi** [comment]
> Do I see it right, that compile always triggers the generation? I'd like it to be opt out

I assume you want this since codegen could take a lot of time, and you wouldn't want to waste it on every incremental compilation?
If that's the case, this is covered by sbt task caching mechanism I added.
If there's no change in openapi files, then there's no need to re-generate.

Is there any other reason I didn't think of?

File: build.sbt
Comment (hochgi) [reply build.sbt]: @987Nabil turns out it works. but only under `++2.12.x`.
Which is expected. When cross version is set to default 2.13, and we try to resolve only 2.12 modules it fails and resorts to trying to download.

So this should work, and we only need carful CI handling for only building/testing the plugin under ++2.12.x

**Review by 987Nabil — APPROVED** [review]

**Pull request merged**
